### PR TITLE
feat: Add plulist.v2 schema

### PR DESCRIPTION
**Pull request by msilfver**

### CODE REVIEW

File: customer-config/checkout-config/che.plulist.v2.json [hunk 1]
```diff
@@ -0,0 +1,221 @@
...
+    },
+    "PluItemType": {
+      "description": "This is an entry for a specific item, and is placed inside a PLU list.",
+      "type": "object",
+      "required": [
+        "identifier",
+        "images",
+        "type"
+      ],
+      "properties": {
```
Comment (DStadnyk) [inline customer-config/checkout-config/che.plulist.v2.json:101]: We also need an `id` property for the `PluItemType`, so that CCC can apply PATCH operation to it. Otherwise, we'll not be able to merge this field.

The `id` field will simply duplicate the `identifier` field and can be ignored by the consumers

File: customer-config/checkout-config/che.plulist.v2.json [hunk 2]
```diff
@@ -0,0 +1,221 @@
...
+        "images": {
```
Comment (DStadnyk) [inline customer-config/checkout-config/che.plulist.v2.json:102]: We agreed to remove `images` from the PLU Item since it can be fetched from the Items API.
I don't think PLU Item Image should be a part of Customer Controlled Configuration, there's already a service responsible for that.

File: customer-config/checkout-config/che.plulist.v2.json [hunk 3]
```diff
@@ -0,0 +1,221 @@
...
+          ]
+        }
+      }
+    },
+    "PluGroupType": {
+      "description": "This is a PLU - it contains a set of other PLUs to form a hierarchy.",
+      "type": "object",
+      "required": [
+        "id",
+        "images",
```
Comment (DStadnyk) [inline customer-config/checkout-config/che.plulist.v2.json:125]: Should the `images` property be required?
We don't have support for it on the UI yet. I would consider making it optional

File: customer-config/checkout-config/che.plulist.v2.json [hunk 4]
```diff
@@ -0,0 +1,221 @@
...
+            "UNKNOWN"
+          ]
+        }
+      }
+    },
+    "ImagesType": {
+      "type": "array",
+      "items": {
+        "type": "object",
+        "properties": {
```
Comment (DStadnyk) [inline customer-config/checkout-config/che.plulist.v2.json:209]: the `url` property should be required for the image, but the images may be an optional field, giving us more control and flexibility.
I think consumers would expect to have a valid URL of the image, if it's in the schema

**Review by DStadnyk — COMMENTED** [review]
Left some notes

File: customer-config/checkout-config/che.plulist.v2.json (hunk 2 above)
Comment (joakimrylander) [reply customer-config/checkout-config/che.plulist.v2.json:102]: But if the customers have not added an image url to the item data but still wants an image in the PLU List?

File: customer-config/checkout-config/che.plulist.v2.json [hunk 5]
```diff
@@ -0,0 +1,221 @@
...
+    "required": [
+      "id",
+      "plu"
+    ],
+    "properties": {
+      "effectiveDate": {
+        "description": "The date when this PLU list should be applied, if present.",
+        "type": "string",
+        "format": "date",
+        "example": "2021-01-01"
```
Comment (vadimb) [inline customer-config/checkout-config/che.plulist.v2.json:17]: Dates are hard - being we have stores that spans multiple timezone in one contry should we make sure to include it too ?

File: customer-config/checkout-config/che.plulist.v2.json (hunk 1 above)
Comment (msilfver) [reply customer-config/checkout-config/che.plulist.v2.json:101]: This is confusing. Why not just use the identifier?

File: customer-config/checkout-config/che.plulist.v2.json (hunk 2 above)
Comment (msilfver) [reply customer-config/checkout-config/che.plulist.v2.json:102]: I agree with Joakim. If there's an image provided here, that can be used, otherwise fetch from the Item API

File: customer-config/checkout-config/che.plulist.v2.json (hunk 5 above)
Comment (msilfver) [reply customer-config/checkout-config/che.plulist.v2.json:17]: The json schema does not specify the timezone for dates, only date-time. We could add a comment that the date should be considered a UTC date.

File: customer-config/checkout-config/che.plulist.v2.json [hunk 6]
```diff
@@ -0,0 +1,221 @@
...
+        "example": "PLU List"
+      },
+      "description": {
+        "description": "The PLU List description",
+        "type": "string",
+        "minLength": 1,
+        "maxLength": 255,
+        "example": "The description for PLU List"
+      },
+      "deviceTypes": {
```
Comment (vadimb) [inline customer-config/checkout-config/che.plulist.v2.json:40]: Is this per category or per entire PLU? We need it to be per catergory so that a till can fetch SCO or POS categorie

File: customer-config/checkout-config/che.plulist.v2.json (hunk 1 above)
Comment (DmitriyStadnik) [reply customer-config/checkout-config/che.plulist.v2.json:101]: We need both, otherwise, CCC API wouldn't be able to PATCH PLU lists. Everyone except for the frontend can ignore it.
FYI @mdovhopo

File: customer-config/checkout-config/che.plulist.v2.json [hunk 7]
```diff
@@ -0,0 +1,221 @@
+{
+  "$schema": "https://json-schema.org/draft/2020-12/schema",
+  "$id": "https://raw.githubusercontent.com/extenda/hiiretail-json-schema-registry/master/customer-config/checkout-config/che.plulist.v1.json",
+  "title": "PLU List messages",
+  "type": "array",
+  "items": {
+    "type": "object",
+    "required": [
+      "id",
+      "plu"
```
Comment (nkv843) [inline customer-config/checkout-config/che.plulist.v2.json:10]: content instead of plu?

File: customer-config/checkout-config/che.plulist.v2.json (hunk 3 above)
Comment (nkv843) [reply customer-config/checkout-config/che.plulist.v2.json:125]: also lets make group content required

File: customer-config/checkout-config/che.plulist.v2.json (hunk 6 above)
Comment (nkv843) [reply customer-config/checkout-config/che.plulist.v2.json:40]: so, in this case, instead of an array of lists we can have a single value, and can we change type at line 5 from array to object?

File: customer-config/checkout-config/che.plulist.v2.json [hunk 8]
```diff
@@ -0,0 +1,221 @@
...
+        "description": "The PLU List description",
+        "type": "string",
+        "minLength": 1,
+        "maxLength": 255,
+        "example": "The description for PLU List"
+      },
+      "deviceTypes": {
+        "description": "A list of device types that this PLUList applies to",
+        "type": "array",
+        "example": "[{'id': 'SelfCheckout'}]",
```
Comment (nkv843) [inline customer-config/checkout-config/che.plulist.v2.json:43]: lets define fixed list of such ids and try to stick to it
on our end we have now 4 types:

PLU_BAGS_SCO
PLU_MPOS
PLU_SCO
PLU_POS

if that's acceptable, let's keep them that way?

File: customer-config/checkout-config/che.plulist.v2.json (hunk 3 above)
Comment (nkv843) [reply customer-config/checkout-config/che.plulist.v2.json:125]: as well as name

File: customer-config/checkout-config/che.plulist.v2.json (hunk 1 above)
Comment (msilfver) [reply customer-config/checkout-config/che.plulist.v2.json:101]: You should patch using the identifier

File: customer-config/checkout-config/che.plulist.v2.json (hunk 3 above)
Comment (msilfver) [reply customer-config/checkout-config/che.plulist.v2.json]: Changed

File: customer-config/checkout-config/che.plulist.v2.json (hunk 8 above)
Comment (msilfver) [reply customer-config/checkout-config/che.plulist.v2.json]: Added those strings as examples. Is that enough or should we use an enum instead?

**Review by nkv843 — APPROVED** [review]

**Review by vadimb — APPROVED** [review]

File: customer-config/checkout-config/che.plulist.v2.json (hunk 1 above)
Comment (mdovhopo) [reply customer-config/checkout-config/che.plulist.v2.json:104]: @msilfver this is pretty simple. we decided to use `id` as a merging key.

check out our adr - https://github.com/extenda/engineering-cloud-core-common/blob/master/docs/customer-controlled-configuration/internal/adr/0008-add-patch-operation-for-all-targets.md

This is a limitation that we agreed to have to simplify delivery on our side and make sure we deliver plu on time.

tldr: we need some field to use as merging key, so we can add elements to array easily. Since you do not want to use `id` instead of `identifier` as we asked, we though of compromise with @DmitriyStadnik:
* We will keep `identifier` field for you, so you do not have to change anything.
* We will have `id` in every element of any array, achieving unified name. this will be used by frontend as unique anchor.
* You will not need to use `id` if you do not want
* `id` will be always === `identifier`, to keep compatibility

File: customer-config/checkout-config/che.plulist.v2.json (hunk 8 above)
Comment (mdovhopo) [reply customer-config/checkout-config/che.plulist.v2.json]: i'd use a string, to simplify future addition of new types. if we go with enums, we will have to update scheme here every time

**Review by mdovhopo — APPROVED** [review]

File: customer-config/checkout-config/che.plulist.v2.json (hunk 1 above)
Comment (glebbash) [reply customer-config/checkout-config/che.plulist.v2.json:104]: @msilfver Is there any reason why we can't use object instead of array on top level?

It would really simplify logic in CCC if it was object.

Current approach with array and `id` field:

```js
[
  { id: "plu1", content: { ... } },
  { id: "plu2", content: { ... } },
]
```

Object on top level where `id` field just become object keys:

```js
{
  plu1: { ... },
  plu2: { ... },
}
```

Comment (mdovhopo) [reply customer-config/checkout-config/che.plulist.v2.json:104]: i guess that would lead to a lot of changes on SD side... they want to do as little changes as possible...

Comment (msilfver) [reply customer-config/checkout-config/che.plulist.v2.json:103]: @glebbash No, it was just as it was before and as I assume most other schemas would look like.

Comment (mdovhopo) [reply customer-config/checkout-config/che.plulist.v2.json:103]: as a compromise we offer having both fields. But i (and FE devs) would prefer to have only `id` for consistency

File: customer-config/checkout-config/che.plulist.v2.json [hunk 9]
```diff
@@ -0,0 +1,227 @@
+{
+  "$schema": "https://json-schema.org/draft/2020-12/schema",
+  "$id": "https://raw.githubusercontent.com/extenda/hiiretail-json-schema-registry/master/customer-config/checkout-config/che.plulist.v1.json",
```
Comment (mdovhopo) [inline customer-config/checkout-config/che.plulist.v2.json:3]: id is off, it should point to v2

File: customer-config/checkout-config/che.plulist.v2.json (hunk 1 above)
Comment (msilfver) [reply customer-config/checkout-config/che.plulist.v2.json:103]: Both fields is even worse

Comment (glebbash) [reply customer-config/checkout-config/che.plulist.v2.json:103]: How about use object instead of array of objects and then you could keep only identifier but would also use it as object's key

Comment (msilfver) [reply customer-config/checkout-config/che.plulist.v2.json:103]: Clarified the property name to itemIdentifier

Comment (msilfver) [reply customer-config/checkout-config/che.plulist.v2.json:99]: @glebbash Changed to object

**Review by mdovhopo — APPROVED** [review]
